staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go

```diff
@@ -82,3 +96,73 @@ func (r *loopbackResolver) ResolveEndpoint(namespace, name string, port int32) (
 	}
 	return r.delegate.ResolveEndpoint(namespace, name, port)
 }
+
+type fakeLoopbackResolver struct{}
+
+func (f *fakeLoopbackResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+	return nil, errors.New("a lookback resolver hasn't been provided")
+}
+
+// NewExtendedEndpointServiceResolver returns a service resolver that extends ServiceResolver interface by providing a few additional methods for supporting requests retries
+func NewExtendedEndpointServiceResolver(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, loopbackResolver ServiceResolver) ServiceResolver {
+	if loopbackResolver == nil {
+		loopbackResolver = &fakeLoopbackResolver{}
+	}
+	return &serviceResolver{
+		services:         services,
+		endpoints:        endpoints,
+		loopbackResolver: loopbackResolver,
+	}
+}
+
+type serviceResolver struct {
+	services         listersv1.ServiceLister
+	endpoints        listersv1.EndpointsLister
+	loopbackResolver ServiceResolver
+}
+
+// ResolveEndpoint resolves (randomly) an endpoint to a given service.
+//
+// Note: Kube uses one service resolver for webhooks and the aggregator this method satisfies webhook.ServiceResolver interface
+func (r *serviceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+	localEndpoint, err := r.loopbackResolver.ResolveEndpoint(namespace, name, port)
+	if err == nil && localEndpoint != nil {
+		return localEndpoint, nil
+	}
+
+	return proxy.ResolveEndpoint(r.services, r.endpoints, namespace, name, port)
+}
+
+// ResolveEndpointWithVisited resolves an endpoint excluding already visited ones. Facilitates supporting retry mechanisms.
+func (r *serviceResolver) ResolveEndpointWithVisited(namespace, name string, port int32, visitedEPs []*url.URL) (*url.URL, error) {
+	localEndpoint, err := r.loopbackResolver.ResolveEndpoint(namespace, name, port)
+	if err == nil && localEndpoint != nil {
+		return localEndpoint, nil
+	}
+
+	return proxy.ResolveEndpoint(r.services, r.endpoints, namespace, name, port, visitedEPs...)
+}
+
+// EndpointCount return the total number of endpoints for the given service
+func (r *serviceResolver) EndpointCount(namespace, name string, port int32) (int, error) {
+	localEndpoint, err := r.loopbackResolver.ResolveEndpoint(namespace, name, port)
+	if err == nil && localEndpoint != nil {
+		return 1, nil
+	}
+
+	allEndpoints := []*url.URL{}
+
+	for {
+		newEndpoint, err := proxy.ResolveEndpoint(r.services, r.endpoints, namespace, name, port, allEndpoints...)
+		if err != nil {
+			break
+		}
+		allEndpoints = append(allEndpoints, newEndpoint)
```

Comment:
this is kind of odd to probe the ResolveEndpoint method n times to find out that it has n endpoints

Reply:
Would expect something more direct.